File: api/rendezvous.py
from __future__ import annotations

import threading
import time
from typing import Dict, List, Literal, Optional, Tuple

from fastapi import FastAPI, Query, Request
from pydantic import BaseModel, Field, field_validator


TransportType = Literal["tcp", "ws"]
FamilyType = Literal["ipv4", "ipv6"]
NatType = Literal["unknown", "open", "restricted", "symmetric"]


class Endpoint(BaseModel):
    transport: TransportType
    host: str = Field(min_length=1, max_length=255)
    port: int = Field(ge=1, le=65535)
    family: FamilyType = "ipv4"
    nat: NatType = "unknown"


class Capabilities(BaseModel):
    relay: bool = False
    transport: List[TransportType] = Field(default_factory=lambda: ["tcp"])


class RegisterRequest(BaseModel):
    node_id: str = Field(min_length=8, max_length=128)
    circle_hint: str = Field(min_length=8, max_length=128)
    endpoints: List[Endpoint] = Field(default_factory=list)
    capabilities: Capabilities = Field(default_factory=Capabilities)
    ttl_s: int = Field(default=120, ge=30, le=600)

    @field_validator("endpoints")
    @classmethod
    def cap_endpoints(cls, value: List[Endpoint]) -> List[Endpoint]:
        if len(value) > 16:
            raise ValueError("too many endpoints")
        return value


class UnregisterRequest(BaseModel):
    node_id: str = Field(min_length=8, max_length=128)
    circle_hint: str = Field(min_length=8, max_length=128)


class PresenceRecord(BaseModel):
    node_id: str
    circle_hint: str
    endpoints: List[Endpoint]
    capabilities: Capabilities
    observed_at: int
    expires_at: int


class RegisterResponse(BaseModel):
    ok: bool = True
    server_time: int
    expires_at: int
    observed_endpoint: Optional[Endpoint] = None


class PeersResponse(BaseModel):
    ok: bool = True
    peers: List[PresenceRecord]


class HealthResponse(BaseModel):
    ok: bool = True
    version: str
    time: int


class PresenceStore:
    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._records: Dict[Tuple[str, str], PresenceRecord] = {}

    def upsert(
        self,
        req: RegisterRequest,
        observed_host: Optional[str],
        observed_port: Optional[int],
    ) -> PresenceRecord:
        now = int(time.time())
        expires_at = now + req.ttl_s

        with self._lock:
            self._cleanup_locked(now)
            record = PresenceRecord(
                node_id=req.node_id,
                circle_hint=req.circle_hint,
                endpoints=req.endpoints,
                capabilities=req.capabilities,
                observed_at=now,
                expires_at=expires_at,
            )
            self._records[(req.circle_hint, req.node_id)] = record
        return record

    def remove(self, req: UnregisterRequest) -> bool:
        with self._lock:
            return self._records.pop((req.circle_hint, req.node_id), None) is not None

    def list_peers(self, circle_hint: str, exclude_node_id: Optional[str], limit: int) -> List[PresenceRecord]:
        now = int(time.time())
        with self._lock:
            self._cleanup_locked(now)
            peers = [
                record
                for (record_circle, _), record in self._records.items()
                if record_circle == circle_hint
            ]

        if exclude_node_id:
            peers = [record for record in peers if record.node_id != exclude_node_id]

        peers.sort(key=lambda record: record.observed_at, reverse=True)
        return peers[:limit]

    def _cleanup_locked(self, now: int) -> None:
        expired_keys = [key for key, record in self._records.items() if record.expires_at <= now]
        for key in expired_keys:
            self._records.pop(key, None)


app = FastAPI(title="Felund Rendezvous API", version="0.1.0")
store = PresenceStore()


@app.get("/v1/health", response_model=HealthResponse)
async def health() -> HealthResponse:
    return HealthResponse(version="0.1.0", time=int(time.time()))


@app.post("/v1/register", response_model=RegisterResponse)
async def register(req: RegisterRequest, request: Request) -> RegisterResponse:
    observed_host = request.client.host if request.client else None
    observed_port = request.client.port if request.client else None
    record = store.upsert(req, observed_host, observed_port)

    observed_endpoint = None
    if observed_host and observed_port:
        observed_endpoint = Endpoint(
            transport="tcp",
            host=observed_host,
            port=observed_port,
            family="ipv4" if ":" not in observed_host else "ipv6",
            nat="unknown",
        )

    return RegisterResponse(
        server_time=record.observed_at,
        expires_at=record.expires_at,
        observed_endpoint=observed_endpoint,
    )


@app.get("/v1/peers", response_model=PeersResponse)
async def peers(
    circle_hint: str = Query(min_length=8, max_length=128),
    limit: int = Query(default=50, ge=1, le=200),
    request: Request = None,
) -> PeersResponse:
    caller_node = request.headers.get("X-Felund-Node") if request else None
    records = store.list_peers(circle_hint=circle_hint, exclude_node_id=caller_node, limit=limit)
    return PeersResponse(peers=records)


@app.delete("/v1/register")
async def unregister(req: UnregisterRequest) -> dict:
    removed = store.remove(req)
    return {"ok": True, "removed": removed}

File: chat-webclient/src/network/rendezvous.ts
import { hmacHex, sha256Hex } from '../core/crypto'

export interface RendezvousHealth {
  ok: boolean
  version?: string
  time?: number
}

export interface RendezvousPeer {
  node_id: string
  circle_hint: string
  endpoints: Array<{
    transport: 'tcp' | 'ws'
    host: string
    port: number
    family: 'ipv4' | 'ipv6'
    nat: 'unknown' | 'open' | 'restricted' | 'symmetric'
  }>
  capabilities: {
    relay: boolean
    transport: Array<'tcp' | 'ws'>
  }
  observed_at: number
  expires_at: number
}

export const normalizeRendezvousBase = (raw: string): string => raw.trim().replace(/\/+$/, '')

const withV1 = (base: string, path: string): string => {
  const normalized = normalizeRendezvousBase(base)
  if (normalized.endsWith('/v1')) {
    return `${normalized}/${path}`
  }
  return `${normalized}/v1/${path}`
}

const circleHint = async (circleId: string): Promise<string> => {
  const digest = await sha256Hex(circleId)
  return digest.slice(0, 16)
}

// ── Request signing ───────────────────────────────────────────────────────────

/**
 * Derive the API signing key: HMAC-SHA256(circle_secret_bytes, "api-v1").
 *
 * The signing key is a one-way transform of the circle secret. It is safe to
 * send to the server during registration — the circle secret itself is never
 * transmitted.
 */
const signingKeyHex = async (circleSecretHex: string): Promise<string> =>
  hmacHex(circleSecretHex, 'api-v1')

/**
 * Return a hex HMAC-SHA256 signature for a canonical API request.
 *
 * Canonical string: METHOD + path + sha256(bodyStr) + ts + nonce
 * Signing key:      HMAC-SHA256(circle_secret_bytes, "api-v1")
 */
const signRequest = async (
  circleSecretHex: string,
  method: string,
  path: string,
  bodyStr: string,
  ts: number,
  nonce: string,
): Promise<string> => {
  const key = await signingKeyHex(circleSecretHex)
  const bodyHash = await sha256Hex(bodyStr)
  const canonical = `${method.toUpperCase()}${path}${bodyHash}${ts}${nonce}`
  return hmacHex(key, canonical)
}

const buildAuthHeaders = async (
  circleSecretHex: string,
  nodeId: string,
  method: string,
  path: string,
  bodyStr: string,
): Promise<Record<string, string>> => {
  const ts = Math.floor(Date.now() / 1000)
  const nonce = Array.from(crypto.getRandomValues(new Uint8Array(16)))
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('')
  const sig = await signRequest(circleSecretHex, method, path, bodyStr, ts, nonce)
  return {
    'X-Felund-Node': nodeId,
    'X-Felund-Ts': String(ts),
    'X-Felund-Nonce': nonce,
    'X-Felund-Signature': sig,
  }
}

// ── Public API ────────────────────────────────────────────────────────────────

export const healthCheck = async (base: string): Promise<RendezvousHealth> => {
  const url = withV1(base, 'health')
  const response = await fetch(url)
  if (!response.ok) {
    throw new Error(`Health check failed (${response.status})`)
  }
  return (await response.json()) as RendezvousHealth
}

export const lookupPeers = async (
  base: string,
  nodeId: string,
  circleId: string,
  limit = 20,
  circleSecretHex = '',
): Promise<RendezvousPeer[]> => {
  const hint = await circleHint(circleId)
  const query = new URLSearchParams({ circle_hint: hint, limit: String(limit) })
  const url = `${withV1(base, 'peers')}?${query.toString()}`

  const baseHeaders: Record<string, string> = { 'X-Felund-Node': nodeId }
  const authHeaders = circleSecretHex
    ? await buildAuthHeaders(circleSecretHex, nodeId, 'GET', '/peers', '')
    : baseHeaders

  const response = await fetch(url, { headers: authHeaders })
  if (!response.ok) {
    throw new Error(`Peer lookup failed (${response.status})`)
  }

  const payload = (await response.json()) as { ok: boolean; peers?: RendezvousPeer[] }
  return payload.peers ?? []
}

const defaultEndpoint = () => {
  const host = window.location.hostname || 'localhost'
  const inferredPort = window.location.protocol === 'https:' ? 443 : 80
  const port = Number.parseInt(window.location.port || '', 10) || inferredPort
  return {
    transport: 'ws' as const,
    host,
    port,
    family: host.includes(':') ? ('ipv6' as const) : ('ipv4' as const),
    nat: 'unknown' as const,
  }
}

export const registerPresence = async (
  base: string,
  input: { nodeId: string; circleId: string; ttlS?: number; circleSecretHex?: string },
): Promise<void> => {
  const hint = await circleHint(input.circleId)
  const body: Record<string, unknown> = {
    node_id: input.nodeId,
    circle_hint: hint,
    endpoints: [defaultEndpoint()],
    capabilities: { relay: false, transport: ['ws'], can_anchor: false },
    ttl_s: input.ttlS ?? 120,
  }

  let authHeaders: Record<string, string> = {
    'content-type': 'application/json',
    'X-Felund-Node': input.nodeId,
  }

  if (input.circleSecretHex) {
    // Include the signing key so the server can store and verify future requests.
    body.signing_key = await signingKeyHex(input.circleSecretHex)
    const bodyStr = JSON.stringify(body)
    const signed = await buildAuthHeaders(
      input.circleSecretHex,
      input.nodeId,
      'POST',
      '/register',
      bodyStr,
    )
    authHeaders = { 'content-type': 'application/json', ...signed }
  }

  const response = await fetch(withV1(base, 'register'), {
    method: 'POST',
    headers: authHeaders,
    body: JSON.stringify(body),
  })

  if (!response.ok) {
    throw new Error(`Register failed (${response.status})`)
  }
}

export const unregisterPresence = async (
  base: string,
  input: { nodeId: string; circleId: string; circleSecretHex?: string },
): Promise<void> => {
  const hint = await circleHint(input.circleId)
  const body = {
    node_id: input.nodeId,
    circle_hint: hint,
  }

  let authHeaders: Record<string, string> = { 'content-type': 'application/json' }

  if (input.circleSecretHex) {
    const bodyStr = JSON.stringify(body)
    const signed = await buildAuthHeaders(
      input.circleSecretHex,
      input.nodeId,
      'DELETE',
      '/register',
      bodyStr,
    )
    authHeaders = { 'content-type': 'application/json', ...signed }
  }

  const response = await fetch(withV1(base, 'register'), {
    method: 'DELETE',
    headers: authHeaders,
    keepalive: true,
    body: JSON.stringify(body),
  })
  if (!response.ok) {
    throw new Error(`Unregister failed (${response.status})`)
  }
}
